### ISSUE TITLE
Allow session.run to execute something in the background or in parallel

### ISSUE THREAD

**Issue by AGeekInside** [theacodes/nox#198]
**How would this feature be useful?**
I am using nox as a way to do some local integration style testing of a service that processes messages. My current test flow is:
1. Start message processor
2. While message processor is running, start data generator
3. Verify that the message processor processed all the generated data

**Describe the solution you'd like**
It would be nice if there was a way to flag a session.run command to execute in the background. As context manager or something would be helpful.

**Describe alternatives you've considered**
I am probably going to go the multiprocessing route within my noxfile for now, but it'd be nice to have it be baked directly into nox.

**Comment by theacodes** [comment]
I would *love* this feature. Especially for projects that need to do integration tests where they start a webserver and then run tests against it, this would be amazing.

A possible api would be something like:

```python
server = session.background("python", "-m", "webapp")
session.run("pytest", "tests")
server.shutdown(deadline=5)
```

**Comment by AGeekInside** [comment]
I was going to take a stab at this, since it's a need we have. Just curious what approach folks think might work. @theacodes were you thinking session.run would start a separate process or something else?

This seems like it might be something that would be related #88 as well, if it was done correctly.

**Comment by theacodes** [comment]
`session.background` would fork() then execv()

**Comment by theacodes** [comment]
or even just subprocess.Popen() and keep track of the process handle.

**Comment by AGeekInside** [comment]
Just some more thoughts/questions, as I think through the implementation. 

1. Should a background command necessarily run in the same virtualenv and use the same shared env variables? I would assume yes, since it's within the body a given session, but I could see the desire to kick off some background command that might have different python package requirements.

2. Would it make sense to have it reuse the code from the nox.run function? Or at least refactor some of the run function to allow for it to be used to run commands as a separate process vs. a blocking call. 

Generally, I like an option that keeps track of the process handle, so it seems like a separate process could use run(), with some way of keeping track of the process handle. I'm not up to speed on the trade-offs between execv vs. popen, so I'll read up on that a bit more.

**Comment by theacodes** [comment]
(1) yes. (2) maybe, but don't twist that function too much to make it work
for this case.

**Comment by obestwalter** [comment]
Unless you are heading for test framework agnostic functionality, this use case is already well covered by pytest itself.

The envisioned API was:

```python
server = session.background("python", "-m", "webapp")
session.run("pytest", "tests")
server.shutdown(deadline=5)
```

in pytest this would be a session scoped fixture.

```python
import pytest

@pytest.fixture(name="socketAddr", scope="session"):
def manage_webserver():
	# start the webserver in the background
	yield ip, port
	# shutdown the webserver
```

A test using this would look like:

```python
import pytest

def test_some_web_thing(socketAddr):
	...
```

Dealing with things like these directly in pytest (or some other test framework) has several advantages:
* The decision if the web server needs to be started for a particular testrun is automatically correct as the server is only started if tests are run depending on the fixture
* information that might change can be passed directly to the test (like a potentially changing socket address)
* the test is not concerned with what it is dealing with - it just asks for the socket address and this might be a local test server or a live server, which means I can reuse those tests in different scenarios
* To understand how the tests work I do not have to understand what nox and my test framework does (and how they interact) because it all happens in the test framework

There is also a plugin (mainly for local testing I think) to start processes that should be alive: [pytest-xprocess](https://github.com/pytest-dev/pytest-xprocess/)

**Comment by AGeekInside** [comment]
Thanks for the pointer to the pytest functionality. I hadn't seen it before. It'll definitely help me get going with our internal needs, since we are using pytest already. 

I do have a fork where I was looking at how to add this to nox based off the comments above, so I'm planning to work on a pull request as I have time.

**Comment by staticdev** [comment]
@obestwalter interesting I just new about [pytest-xdist](https://pypi.org/project/pytest-xdist/).

@theacodes Couldn't session.background just spawn a thread like in here?
https://github.com/tox-dev/tox/blob/master/src/tox/session/commands/run/parallel.py

**Comment by davidism** [comment]
I'm not sure if this would solve #22, which was closed in favor of this. Unless the intention is that one nox session could call a bunch of other nox sessions as background tasks. I think that's an interesting idea, it would definitely keep "run these sessions in parallel, then run this reporting session" simple. But it would still require the author of the session function to handle detecting when the task was finished and handling parallel output and reporting. Could you clarify how this idea will enable parallel sessions?

**Comment by FollowTheProcess** [comment]
Continuing my walk through some older issues.

It sounds to me like there are two things being conflated here:

1) The ability to do something like `session.background` for launching a web server or other background tasks needed by _that session_. This could include calling other sessions but that's not really the main idea from the sounds of it.
2) The ability to do something like `nox --parallel` and have the sessions run concurrently in a threadpool or similar

I think this issue is more closely linked to **1** so I'll leave it open to track this option

I'll open another one to better match point **2**

**Comment by FollowTheProcess** [comment]
Just a thought on the API here, I think a context manager would be the cleanest way of doing this IMO, taking inspiration from how `concurrent.futures` handles things

Something like:

```python
@nox.session
def do_stuff_in_the_background(session):
    with session.background("launch", "a", "webserver"):
        session.run("thing", "that" "depends", "on", "webserver")
    # Now we're back in the main session again, the background task is cleaned up
    # and the session exists normally
```

This has the advantages of:

* Consistency with how a user would think about launching a background process i.e. "I need this thing to run in the background _while_ this other thing happens"
* Consistency with stdlib stuff like `concurrent.futures`
* The context manager ensures cleaning up of resources (spawned subprocesses or threads, however this gets implemented)
* The user doesn't have to worry about doing something like `session.cleanup()` at the end of the session or any weird bugs that might arise from leaving it running unintentionally

**Comment by sneakers-the-rat** [comment]
seems like this is a little stale, but for what it's worth, working on a project that uses nox rn and really want something like [concurrently](https://www.npmjs.com/package/concurrently), so since usually there's not *that* much of a difference between "foreground" and "background" tasks except who gets stdout, i think something like this would be a p natural syntax without the awkwardness of needing to nest a bunch of context managers or have the semantic ambiguity of the singularness/pluralness of "background" if one wanted multiple processes (that each do a different thing, ie. != running same thing n times).

```python
@nox.session
def concurrent_thing(session):
    session.concurrently(
        ["sphinx-autobuild", "-D", "language='en'", ".", "_build/en"],
        ["sphinx-autobuild", "-D", "language='es'", ".", "_build/es"],
        ["sphinx-autobuild", "-D", "language='jp'", ".", "_build/jp"],
        # ...
    )
```

and, to pass params, like `pytest.param()`...

```python
@nox.session
def concurrent_thing(session):
    session.concurrently(
        nox.task(["sphinx-autobuild", "-D", "language='en'", ".", "_build/en"], label="en"),
        nox.task(["sphinx-autobuild", "-D", "language='es'", ".", "_build/es"], label="es"),
        nox.task(["sphinx-autobuild", "-D", "language='jp'", ".", "_build/jp"], label="jp"),
        # ...
    )
```

**Comment by alex** [comment]
Another potential variant would be something like:

```python
with session.nursery() as nursery:
    handle1 = nursery.run(["..."])
    handle2 = nursery.run(["..."])
    handle3 = nursery.run(["..."])

output = handle1.resolve()
```

where `nursery` would block until all commands had finished, and then handles could be used to get the buffered output later.

If there's interest in any APIs like this, happy to work up a PR
